File: haskell/src/View/UI.hs
module View.UI where

import System.Console.ANSI
import System.IO (hFlush, stdout)
import Data.Char (isSpace)
import Data.List (dropWhileEnd)

logo :: String
logo =
    "   █████╗ ██╗      ██████╗  ██████╗  █████╗ ██████╗  ██████╗  ██████╗  \n" ++
    "  ██╔══██╗██║     ██╔═══██╗██╔════╝ ██╔══██╗██╔══██╗██╔═══██╗ ██╔══██╗ \n" ++
    "  ███████║██║     ██║   ██║██║      ███████║██   ██║██║   ██║ ██████╔╝ \n" ++
    "  ██╔══██║██║     ██║   ██║██║      ██╔══██║██   ██║██║   ██║ ██╔══██╗ \n" ++
    "  ██║  ██║███████╗╚██████╔╝╚██████╔ ██║  ██║███████╝╚██████╔╝ ██║  ██║ \n" ++
    "  ╚═╝  ╚═╝╚══════╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚═════╝  ╚═════╝  ╚═╝  ╚═╝ "

-- | Função para desenhar cabeçalho com logo e título
-- * title: título da tela
drawHeader :: String -> IO ()
drawHeader title = do
    clearScreen
    setCursorPosition 0 0
    -- | Logo em amarelo
    setSGR [SetColor Foreground Vivid Yellow]
    putStrLn logo
    setSGR [Reset]
    putStrLn ""
    -- | Título da tela em ciano
    setSGR [SetColor Foreground Vivid Cyan]
    putStrLn ("===============  " ++ title ++ "  ===============")
    setSGR [Reset]
    putStrLn ""

-- | Função para desenhar um subtítulo
-- * subtitle: subtítulo da seção
drawSubHeader :: String -> IO ()
drawSubHeader subtitle = do
    setSGR [SetColor Foreground Dull Blue]
    putStrLn (subtitle ++ " ---")
    setSGR [Reset]

-- | Trim simples (sem precisar de reverse duas vezes)
trim :: String -> String
trim = dropWhile isSpace . dropWhileEnd isSpace

-- | Quando a entrada é OBRIGATÓRIA (não aceita vazio)
readLine :: String -> IO String
readLine prompt = do
  putStr prompt
  hFlush stdout
  line <- getLine
  let t = trim line
  if null t
     then do
       putStrLn "Entrada vazia, por favor tente novamente."
       readLine prompt
     else
       return t

-- | Quando a entrada PODE ser vazia (Enter para cancelar/voltar)
readLineMaybe :: String -> IO String
readLineMaybe prompt = do
  putStr prompt
  hFlush stdout
  line <- getLine
  return (trim line)

File: prolog/src/interfaces/UI.pl
ansi_clear :- write('\e[2J').
ansi_home :- write('\e[H').
ansi_color(Color) :- format('\e[~wm', [Color]).
ansi_reset :- write('\e[0m').

color_yellow :- ansi_color(33).
color_cyan   :- ansi_color(36).
color_blue   :- ansi_color(34).

logo :-
    color_yellow,
    write('   █████╗ ██╗      ██████╗  ██████╗  █████╗ ██████╗  ██████╗  ██████╗  \n'),
    write('  ██╔══██╗██║     ██╔═══██╗██╔════╝ ██╔══██╗██╔══██╗██╔═══██╗ ██╔══██╗ \n'),
    write('  ███████║██║     ██║   ██║██║      ███████║██   ██║██║   ██║ ██████╔╝ \n'),
    write('  ██╔══██║██║     ██║   ██║██║      ██╔══██║██   ██║██║   ██║ ██╔══██╗ \n'),
    write('  ██║  ██║███████╗╚██████╔╝╚██████╔ ██║  ██║███████╝╚██████╔╝ ██║  ██║ \n'),
    write('  ╚═╝  ╚═╝╚══════╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚═════╝  ╚═════╝  ╚═╝  ╚═╝ \n'),
    ansi_reset.

draw_header(Title) :-
    ansi_clear, ansi_home,
    logo, nl,
    color_cyan,
    format("===============  ~w  ===============~n", [Title]),
    ansi_reset, nl.

draw_subheader(Sub) :-
    color_blue,
    format("~w ---~n", [Sub]),
    ansi_reset.

trim(Str, Trimmed) :-
    atom_string(A, Str),
    normalize_space(atom(T), A),
    atom_string(T, Trimmed).

read_line(Prompt, Line) :-
    write(Prompt), flush_output,
    read_line_to_string(user_input, Input),
    trim(Input, Trimmed),
    ( Trimmed = "" ->
        write('Entrada vazia, tente novamente.\n'),
        read_line(Prompt, Line)
    ; Line = Trimmed
    ).

read_line_maybe(Prompt, Line) :-
    write(Prompt), flush_output,
    read_line_to_string(user_input, Input),
    trim(Input, Line).
